File: src/tq/reporting/json.py
"""Machine-readable JSON reporting for tq diagnostics."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path

    from rich.console import Console

    from tq.engine.models import EngineResult, Finding


def print_json_report(*, result: EngineResult, console: Console, cwd: Path) -> None:
    """Print deterministic JSON diagnostics payload.

    Args:
        result: Engine result to serialize.
        console: Rich console to write output to.
        cwd: Path used to relativize diagnostic paths.
    """
    payload = {
        "findings": [
            _finding_payload(finding=finding, cwd=cwd) for finding in result.findings
        ],
        "summary": {
            "errors": result.summary.errors,
            "warnings": result.summary.warnings,
            "infos": result.summary.infos,
            "total": result.summary.total,
        },
    }
    serialized = json.dumps(payload, ensure_ascii=False, separators=(",", ":"))
    console.file.write(f"{serialized}\n")


def _finding_payload(*, finding: Finding, cwd: Path) -> dict[str, str | int | None]:
    """Convert a finding into the stable JSON representation."""
    return {
        "rule_id": finding.rule_id.value,
        "severity": finding.severity.value,
        "message": finding.message,
        "path": _display_path(path=finding.path, cwd=cwd),
        "line": finding.line,
        "suggestion": finding.suggestion,
        "target": finding.target,
    }


def _display_path(*, path: Path, cwd: Path) -> str:
    """Render a path relative to cwd when possible."""
    try:
        return path.relative_to(cwd).as_posix()
    except ValueError:
        return path.as_posix()

File: crates/tq-reporting/src/json.rs
use std::io::Write;
use std::path::Path;

use serde::Serialize;
use tq_engine::{EngineResult, Finding};

use crate::ReportingError;
use crate::path::display_path;

pub struct JsonReporter<'a> {
    cwd: &'a Path,
}

impl<'a> JsonReporter<'a> {
    #[must_use]
    pub const fn new(cwd: &'a Path) -> Self {
        Self { cwd }
    }

    pub fn write<W: Write>(
        &self,
        writer: &mut W,
        result: &EngineResult,
    ) -> Result<(), ReportingError> {
        let payload = JsonReport {
            findings: result
                .findings()
                .iter()
                .map(|finding| JsonFinding::from_finding(finding, self.cwd))
                .collect(),
            summary: JsonSummary::from_result(result),
        };

        serde_json::to_writer(&mut *writer, &payload)?;
        writer.write_all(b"\n")?;
        Ok(())
    }
}

#[derive(Serialize)]
struct JsonReport<'a> {
    findings: Vec<JsonFinding<'a>>,
    summary: JsonSummary,
}

#[derive(Serialize)]
struct JsonFinding<'a> {
    rule_id: &'a str,
    severity: &'a str,
    message: &'a str,
    path: String,
    line: Option<u32>,
    suggestion: Option<&'a str>,
    target: Option<&'a str>,
}

impl<'a> JsonFinding<'a> {
    fn from_finding(finding: &'a Finding, cwd: &Path) -> Self {
        Self {
            rule_id: finding.rule_id().as_str(),
            severity: finding.severity().as_str(),
            message: finding.message(),
            path: display_path(finding.path(), cwd),
            line: finding.line(),
            suggestion: finding.suggestion(),
            target: finding.target(),
        }
    }
}

#[derive(Serialize)]
struct JsonSummary {
    errors: usize,
    warnings: usize,
    infos: usize,
    total: usize,
}

impl JsonSummary {
    const fn from_result(result: &EngineResult) -> Self {
        let summary = result.summary();
        Self {
            errors: summary.errors(),
            warnings: summary.warnings(),
            infos: summary.infos(),
            total: summary.total(),
        }
    }
}
